### PR TITLE
[DO NOT MERGE] Org features/permissions refactor

### PR DESCRIPTION
**Pull request by ryanpbrewster**
So...this PR might be kind of a disaster. Part of it feels awesome to me, and the rest of it is hot garbage.

@leogrim I think there is a better way to handle the types here. Can you apply some of your type wizardry?

@camhashemi I'm so glad that you wrote good tests for this stuff. It made this refactor like a million times easier.

@andrewconner and @stkem do you have any input on the architecture here?

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    case object DISABLED extends FeatureSetting[ExportKeeps.type]("disabled")
+    case object ADMINS extends FeatureSetting[ExportKeeps.type]("admins")
+    case object MEMBERS extends FeatureSetting[ExportKeeps.type]("members")
+    def targetRoles(setting: FeatureSetting[ExportKeeps.type]): Set[Option[OrganizationRole]] = setting match {
+      case DISABLED => Set.empty
+      case ADMINS => Set(Some(OrganizationRole.ADMIN))
+      case MEMBERS => Set(Some(OrganizationRole.ADMIN), Some(OrganizationRole.MEMBER))
+    }
+  }
+}
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala:233]: This is whatever the opposite of DRY is. WET? "Whatever, Enumerate Things"?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationCommander.scala [hunk 2]
```diff
@@ -220,20 +221,14 @@ class OrganizationCommanderImpl @Inject() (
...
       else "https://" + url
     }
     val badSiteUrl = normalizedSiteUrl.exists(URI.parse(_).isFailure)
-    (badName, badPermissionsChange, badSiteUrl) match {
-      case (true, _, _) => Some(OrganizationFail.INVALID_MODIFY_NAME)
-      case (_, true, _) => Some(OrganizationFail.INVALID_MODIFY_PERMISSIONS)
-      case (_, _, true) => Some(OrganizationFail.INVALID_MODIFY_SITEURL)
+    (badName, badSiteUrl) match {
+      case (true, _) => Some(OrganizationFail.INVALID_MODIFY_NAME)
+      case (_, true) => Some(OrganizationFail.INVALID_MODIFY_PERMISSIONS)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/OrganizationCommander.scala:231]: Oops, this should be `INVALID_MODIFY_SITEURL`

File: kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+  implicit val format: Format[Feature] = Format (
+    __.read[String].map(Feature(_)),
+    Writes { x => JsString(x.value) }
+  )
+
+  case object PublishLibraries extends OrganizationPermissionFeature(OrganizationPermission.PUBLISH_LIBRARIES)
+  object PublishLibrariesSetting {
+    val format: Format[FeatureSetting[PublishLibraries.type]] = Format(__.read[String].map(toSetting), Writes { x => JsString(x.value) } )
+    def toSetting = Set(DISABLED, ADMINS, MEMBERS).map(x => x.value -> x).toMap.apply _
```
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala:109]: might be handy to have a `.ALL`

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala:109]: This was such a good call. Led to much cleaner code (as you will see soon, I hope).

**Pull request closed without merging**